waku/v2/protocol/peer_exchange/shard_lru.go

```diff
@@ -0,0 +1,137 @@
+	idToNode   map[enode.ID][]*list.Element
+	shardNodes map[ClusterIndex]*list.List
+}
+
+func newShardLRU(size int) *shardLRU {
+	return &shardLRU{
+		idToNode:   map[enode.ID][]*list.Element{},
+		shardNodes: map[ClusterIndex]*list.List{},
+		size:       size,
+	}
+}
+
+type nodeWithClusterIndex struct {
+	key  ClusterIndex
+	node *enode.Node
+}
+
+// time complexity: O(number of previous indexes present for node.ID)
+func (l *shardLRU) remove(node *enode.Node) {
+	elements := l.idToNode[node.ID()]
+	for _, element := range elements {
+		key := element.Value.(nodeWithClusterIndex).key
+		l.shardNodes[key].Remove(element)
+	}
+	delete(l.idToNode, node.ID())
+}
+
+// if a node is removed for a list, remove it from idToNode too
+func (l *shardLRU) removeFromIdToNode(ele *list.Element) {
+	nodeID := ele.Value.(nodeWithClusterIndex).node.ID()
+	for ind, entries := range l.idToNode[nodeID] {
+		if entries == ele {
+			l.idToNode[nodeID] = append(l.idToNode[nodeID][:ind], l.idToNode[nodeID][ind+1:]...)
+			break
+		}
+	}
+	if len(l.idToNode[nodeID]) == 0 {
+		delete(l.idToNode, nodeID)
+	}
+}
+
+// time complexity: O(new number of indexes in node's shard)
+func (l *shardLRU) add(node *enode.Node) {
+	shard, err := wenr.RelaySharding(node.Record())
+	if shard == nil || err != nil { // if no shard info, then add to node to Cluster 0, Index 0
+		shard = &protocol.RelayShards{
+			Cluster: 0,
+			Indices: []uint16{0},
+		}
+	}
+	elements := []*list.Element{}
+	for _, index := range shard.Indices {
+		key := ClusterIndex{
+			shard.Cluster,
+			index,
+		}
+		if l.shardNodes[key] == nil {
+			l.shardNodes[key] = list.New()
+		}
+		if l.shardNodes[key].Len() >= l.size {
+			oldest := l.shardNodes[key].Back()
+			l.removeFromIdToNode(oldest)
+			l.shardNodes[key].Remove(oldest)
+		}
+		entry := l.shardNodes[key].PushFront(nodeWithClusterIndex{
+			key:  key,
+			node: node,
+		})
+		elements = append(elements, entry)
+
+	}
+	l.idToNode[node.ID()] = elements
+}
+
+// this will be called when the seq number of node is more than the one in cache
+func (l *shardLRU) Add(node *enode.Node) {
+	// removing bcz previous node might be subscribed to different shards, we need to remove node from those shards
+	l.remove(node)
+	l.add(node)
+}
+
+// clusterIndex is nil when peers for no specific shard are requested
+func (l shardLRU) getNodes(clusterIndex *ClusterIndex) []*enode.Node {
+	// if clusterIndex is nil, then return all nodes
+	if clusterIndex == nil {
+		nodes := make([]*enode.Node, 0, len(l.idToNode))
+		for _, entries := range l.idToNode {
+			nodes = append(nodes, entries[0].Value.(nodeWithClusterIndex).node)
+		}
+		return nodes
+	}
+	//
+	if entries := l.shardNodes[*clusterIndex]; entries != nil && entries.Len() != 0 {
+		nodes := make([]*enode.Node, 0, entries.Len())
+		for ent := entries.Back(); ent != nil; ent = ent.Prev() {
+			nodes = append(nodes, ent.Value.(nodeWithClusterIndex).node)
+		}
+		return nodes
+	}
```

Comment:
Shouldn't the nodes retrieved be moved to the front so that they are not Least used anymore?

Reply:
That i understand, since this is an LRU which means we need to keep track of which nodes are retrieved from the cache so that when it overflows only the least used gets removed. 
Without that, we cannot call it an LRU. Because we will just be overriding the oldest inserted entry.